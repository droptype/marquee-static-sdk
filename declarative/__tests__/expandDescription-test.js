jest.autoMockOff()


describe('expandDescription', () => {
    const makeDescriptionTree       = require('../makeDescriptionTree')
    const expandDescription         = require('../expandDescription')
    const React                     = require('react')
    const { HTMLView, Enumerate }   = require('../Site')

    const r = React.createElement


    it('should expand the description', () => {
        const expanded = expandDescription(
            makeDescriptionTree(
                r(HTMLView, { props: {}, name: 'home', path: '/', component: () => r('div') },
                    r(HTMLView, { props: {}, name: 'about', path: 'about', component: () => r('div') }),
                    r(Enumerate, { items: [{slug: 'a'}, {slug: 'b'}, {slug: 'c'}] },
                        r(HTMLView, { props: s => s, name: 'story_detail', path: s => s.slug, linkKey: s => s, component: () => r('div') })
                    )
                )
            )
        )
        expect(
            expanded
        ).toMatchSnapshot()
        expect(expanded.children.length).toEqual(4)
        expect(expanded.children[0].props.name).toEqual('about')
        expect(expanded.children[1].props.name).toEqual('story_detail')
    })

    it('should throw if given an enumeration without items', () => {
        expect( () => {
            expandDescription(
                makeDescriptionTree(
                    r(HTMLView, { props: {}, name: 'home', path: '/', component: () => r('div') },
                        r(Enumerate, { },
                            r(HTMLView, { props: s => s, name: 'story_detail', path: s => s.slug, linkKey: s => s, component: () => r('div') })
                        )
                    )

                )
            )
        }).toThrow()
    })

    it('should expand nested enumerates', () => {
        const expanded = expandDescription(
            makeDescriptionTree(
                r(HTMLView, { props: {}, name: 'home', path: '/', component: () => r('div') },
                    r(HTMLView, { props: {}, name: 'about', path: 'about', component: () => r('div') }),
                    r(Enumerate, { items: [{slug: 'a'}, {slug: 'b'}, {slug: 'c'}] },
                        r(HTMLView, { props: c => c, name: 'collection_detail', path: c => c.slug, linkKey: c => c, component: () => r('div') },
                            r(Enumerate, { items: () => ([{slug: 'x'}, {slug: 'y'}, {slug: 'z'}]) },
                                r(HTMLView, { props: c => c, name: 'story_detail', path: c => c.slug, linkKey: c => c, component: () => r('div') })
                            )
                        )
                    )
                )
            )
        )
        expect(
            expanded
        ).toMatchSnapshot()
        expect(expanded.children.length).toEqual(4)
        expect(expanded.children[0].props.name).toEqual('about')
        expect(expanded.children[1].props.name).toEqual('collection_detail')
        expect(expanded.children[1].children.length).toEqual(3)
        expect(expanded.children[1].children[0].props.name).toEqual('story_detail')
    })

    it('should omit children of empty enumerate items', () => {
        const expanded = expandDescription(
            makeDescriptionTree(
                r(HTMLView, { props: {}, name: 'home', path: '/', component: () => r('div') },
                    r(HTMLView, { props: {}, name: 'about', path: 'about', component: () => r('div') }),
                    r(Enumerate, { items: [] },
                        r(HTMLView, { props: c => c, name: 'collection_detail', path: c => c.slug, linkKey: c => c, component: () => r('div') })
                    )
                )
            )
        )
        expect(
            expanded
        ).toMatchSnapshot()
        expect(expanded.children.length).toEqual(1)
    })

})