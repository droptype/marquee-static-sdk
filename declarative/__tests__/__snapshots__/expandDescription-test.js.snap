// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expandDescription should expand nested enumerates 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [],
      "enumeration": undefined,
      "evaluated_path": null,
      "gathered_props": null,
      "link_key": null,
      "parent": [Circular],
      "props": Object {
        "component": [Function],
        "fragment": false,
        "name": "about",
        "path": "about",
        "props": Object {},
      },
      "type": [Function],
    },
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "enumeration": EnumerationItem {
            "index": 0,
            "item": Object {
              "slug": "x",
            },
            "items": Array [
              Object {
                "slug": "x",
              },
              Object {
                "slug": "y",
              },
              Object {
                "slug": "z",
              },
            ],
            "next": Object {
              "slug": "y",
            },
            "path": undefined,
            "previous": null,
          },
          "evaluated_path": null,
          "gathered_props": null,
          "link_key": null,
          "parent": [Circular],
          "props": Object {
            "component": [Function],
            "fragment": false,
            "linkKey": [Function],
            "name": "story_detail",
            "path": [Function],
            "props": [Function],
          },
          "type": [Function],
        },
        Object {
          "children": Array [],
          "enumeration": EnumerationItem {
            "index": 1,
            "item": Object {
              "slug": "y",
            },
            "items": Array [
              Object {
                "slug": "x",
              },
              Object {
                "slug": "y",
              },
              Object {
                "slug": "z",
              },
            ],
            "next": Object {
              "slug": "z",
            },
            "path": undefined,
            "previous": Object {
              "slug": "x",
            },
          },
          "evaluated_path": null,
          "gathered_props": null,
          "link_key": null,
          "parent": [Circular],
          "props": Object {
            "component": [Function],
            "fragment": false,
            "linkKey": [Function],
            "name": "story_detail",
            "path": [Function],
            "props": [Function],
          },
          "type": [Function],
        },
        Object {
          "children": Array [],
          "enumeration": EnumerationItem {
            "index": 2,
            "item": Object {
              "slug": "z",
            },
            "items": Array [
              Object {
                "slug": "x",
              },
              Object {
                "slug": "y",
              },
              Object {
                "slug": "z",
              },
            ],
            "next": undefined,
            "path": undefined,
            "previous": Object {
              "slug": "y",
            },
          },
          "evaluated_path": null,
          "gathered_props": null,
          "link_key": null,
          "parent": [Circular],
          "props": Object {
            "component": [Function],
            "fragment": false,
            "linkKey": [Function],
            "name": "story_detail",
            "path": [Function],
            "props": [Function],
          },
          "type": [Function],
        },
      ],
      "enumeration": EnumerationItem {
        "index": 0,
        "item": Object {
          "slug": "a",
        },
        "items": Array [
          Object {
            "slug": "a",
          },
          Object {
            "slug": "b",
          },
          Object {
            "slug": "c",
          },
        ],
        "next": Object {
          "slug": "b",
        },
        "path": undefined,
        "previous": null,
      },
      "evaluated_path": null,
      "gathered_props": null,
      "link_key": null,
      "parent": [Circular],
      "props": Object {
        "component": [Function],
        "fragment": false,
        "linkKey": [Function],
        "name": "collection_detail",
        "path": [Function],
        "props": [Function],
      },
      "type": [Function],
    },
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "enumeration": EnumerationItem {
            "index": 0,
            "item": Object {
              "slug": "x",
            },
            "items": Array [
              Object {
                "slug": "x",
              },
              Object {
                "slug": "y",
              },
              Object {
                "slug": "z",
              },
            ],
            "next": Object {
              "slug": "y",
            },
            "path": undefined,
            "previous": null,
          },
          "evaluated_path": null,
          "gathered_props": null,
          "link_key": null,
          "parent": [Circular],
          "props": Object {
            "component": [Function],
            "fragment": false,
            "linkKey": [Function],
            "name": "story_detail",
            "path": [Function],
            "props": [Function],
          },
          "type": [Function],
        },
        Object {
          "children": Array [],
          "enumeration": EnumerationItem {
            "index": 1,
            "item": Object {
              "slug": "y",
            },
            "items": Array [
              Object {
                "slug": "x",
              },
              Object {
                "slug": "y",
              },
              Object {
                "slug": "z",
              },
            ],
            "next": Object {
              "slug": "z",
            },
            "path": undefined,
            "previous": Object {
              "slug": "x",
            },
          },
          "evaluated_path": null,
          "gathered_props": null,
          "link_key": null,
          "parent": [Circular],
          "props": Object {
            "component": [Function],
            "fragment": false,
            "linkKey": [Function],
            "name": "story_detail",
            "path": [Function],
            "props": [Function],
          },
          "type": [Function],
        },
        Object {
          "children": Array [],
          "enumeration": EnumerationItem {
            "index": 2,
            "item": Object {
              "slug": "z",
            },
            "items": Array [
              Object {
                "slug": "x",
              },
              Object {
                "slug": "y",
              },
              Object {
                "slug": "z",
              },
            ],
            "next": undefined,
            "path": undefined,
            "previous": Object {
              "slug": "y",
            },
          },
          "evaluated_path": null,
          "gathered_props": null,
          "link_key": null,
          "parent": [Circular],
          "props": Object {
            "component": [Function],
            "fragment": false,
            "linkKey": [Function],
            "name": "story_detail",
            "path": [Function],
            "props": [Function],
          },
          "type": [Function],
        },
      ],
      "enumeration": EnumerationItem {
        "index": 1,
        "item": Object {
          "slug": "b",
        },
        "items": Array [
          Object {
            "slug": "a",
          },
          Object {
            "slug": "b",
          },
          Object {
            "slug": "c",
          },
        ],
        "next": Object {
          "slug": "c",
        },
        "path": undefined,
        "previous": Object {
          "slug": "a",
        },
      },
      "evaluated_path": null,
      "gathered_props": null,
      "link_key": null,
      "parent": [Circular],
      "props": Object {
        "component": [Function],
        "fragment": false,
        "linkKey": [Function],
        "name": "collection_detail",
        "path": [Function],
        "props": [Function],
      },
      "type": [Function],
    },
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "enumeration": EnumerationItem {
            "index": 0,
            "item": Object {
              "slug": "x",
            },
            "items": Array [
              Object {
                "slug": "x",
              },
              Object {
                "slug": "y",
              },
              Object {
                "slug": "z",
              },
            ],
            "next": Object {
              "slug": "y",
            },
            "path": undefined,
            "previous": null,
          },
          "evaluated_path": null,
          "gathered_props": null,
          "link_key": null,
          "parent": [Circular],
          "props": Object {
            "component": [Function],
            "fragment": false,
            "linkKey": [Function],
            "name": "story_detail",
            "path": [Function],
            "props": [Function],
          },
          "type": [Function],
        },
        Object {
          "children": Array [],
          "enumeration": EnumerationItem {
            "index": 1,
            "item": Object {
              "slug": "y",
            },
            "items": Array [
              Object {
                "slug": "x",
              },
              Object {
                "slug": "y",
              },
              Object {
                "slug": "z",
              },
            ],
            "next": Object {
              "slug": "z",
            },
            "path": undefined,
            "previous": Object {
              "slug": "x",
            },
          },
          "evaluated_path": null,
          "gathered_props": null,
          "link_key": null,
          "parent": [Circular],
          "props": Object {
            "component": [Function],
            "fragment": false,
            "linkKey": [Function],
            "name": "story_detail",
            "path": [Function],
            "props": [Function],
          },
          "type": [Function],
        },
        Object {
          "children": Array [],
          "enumeration": EnumerationItem {
            "index": 2,
            "item": Object {
              "slug": "z",
            },
            "items": Array [
              Object {
                "slug": "x",
              },
              Object {
                "slug": "y",
              },
              Object {
                "slug": "z",
              },
            ],
            "next": undefined,
            "path": undefined,
            "previous": Object {
              "slug": "y",
            },
          },
          "evaluated_path": null,
          "gathered_props": null,
          "link_key": null,
          "parent": [Circular],
          "props": Object {
            "component": [Function],
            "fragment": false,
            "linkKey": [Function],
            "name": "story_detail",
            "path": [Function],
            "props": [Function],
          },
          "type": [Function],
        },
      ],
      "enumeration": EnumerationItem {
        "index": 2,
        "item": Object {
          "slug": "c",
        },
        "items": Array [
          Object {
            "slug": "a",
          },
          Object {
            "slug": "b",
          },
          Object {
            "slug": "c",
          },
        ],
        "next": undefined,
        "path": undefined,
        "previous": Object {
          "slug": "b",
        },
      },
      "evaluated_path": null,
      "gathered_props": null,
      "link_key": null,
      "parent": [Circular],
      "props": Object {
        "component": [Function],
        "fragment": false,
        "linkKey": [Function],
        "name": "collection_detail",
        "path": [Function],
        "props": [Function],
      },
      "type": [Function],
    },
  ],
  "enumeration": undefined,
  "evaluated_path": null,
  "gathered_props": null,
  "link_key": null,
  "parent": undefined,
  "props": Object {
    "component": [Function],
    "fragment": false,
    "name": "home",
    "path": "/",
    "props": Object {},
  },
  "type": [Function],
}
`;

exports[`expandDescription should expand the description 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [],
      "enumeration": undefined,
      "evaluated_path": null,
      "gathered_props": null,
      "link_key": null,
      "parent": [Circular],
      "props": Object {
        "component": [Function],
        "fragment": false,
        "name": "about",
        "path": "about",
        "props": Object {},
      },
      "type": [Function],
    },
    Object {
      "children": Array [],
      "enumeration": EnumerationItem {
        "index": 0,
        "item": Object {
          "slug": "a",
        },
        "items": Array [
          Object {
            "slug": "a",
          },
          Object {
            "slug": "b",
          },
          Object {
            "slug": "c",
          },
        ],
        "next": Object {
          "slug": "b",
        },
        "path": undefined,
        "previous": null,
      },
      "evaluated_path": null,
      "gathered_props": null,
      "link_key": null,
      "parent": [Circular],
      "props": Object {
        "component": [Function],
        "fragment": false,
        "linkKey": [Function],
        "name": "story_detail",
        "path": [Function],
        "props": [Function],
      },
      "type": [Function],
    },
    Object {
      "children": Array [],
      "enumeration": EnumerationItem {
        "index": 1,
        "item": Object {
          "slug": "b",
        },
        "items": Array [
          Object {
            "slug": "a",
          },
          Object {
            "slug": "b",
          },
          Object {
            "slug": "c",
          },
        ],
        "next": Object {
          "slug": "c",
        },
        "path": undefined,
        "previous": Object {
          "slug": "a",
        },
      },
      "evaluated_path": null,
      "gathered_props": null,
      "link_key": null,
      "parent": [Circular],
      "props": Object {
        "component": [Function],
        "fragment": false,
        "linkKey": [Function],
        "name": "story_detail",
        "path": [Function],
        "props": [Function],
      },
      "type": [Function],
    },
    Object {
      "children": Array [],
      "enumeration": EnumerationItem {
        "index": 2,
        "item": Object {
          "slug": "c",
        },
        "items": Array [
          Object {
            "slug": "a",
          },
          Object {
            "slug": "b",
          },
          Object {
            "slug": "c",
          },
        ],
        "next": undefined,
        "path": undefined,
        "previous": Object {
          "slug": "b",
        },
      },
      "evaluated_path": null,
      "gathered_props": null,
      "link_key": null,
      "parent": [Circular],
      "props": Object {
        "component": [Function],
        "fragment": false,
        "linkKey": [Function],
        "name": "story_detail",
        "path": [Function],
        "props": [Function],
      },
      "type": [Function],
    },
  ],
  "enumeration": undefined,
  "evaluated_path": null,
  "gathered_props": null,
  "link_key": null,
  "parent": undefined,
  "props": Object {
    "component": [Function],
    "fragment": false,
    "name": "home",
    "path": "/",
    "props": Object {},
  },
  "type": [Function],
}
`;

exports[`expandDescription should omit children of empty enumerate items 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [],
      "enumeration": undefined,
      "evaluated_path": null,
      "gathered_props": null,
      "link_key": null,
      "parent": [Circular],
      "props": Object {
        "component": [Function],
        "fragment": false,
        "name": "about",
        "path": "about",
        "props": Object {},
      },
      "type": [Function],
    },
  ],
  "enumeration": undefined,
  "evaluated_path": null,
  "gathered_props": null,
  "link_key": null,
  "parent": undefined,
  "props": Object {
    "component": [Function],
    "fragment": false,
    "name": "home",
    "path": "/",
    "props": Object {},
  },
  "type": [Function],
}
`;
